pkger/clone_resource.go

```diff
@@ -31,6 +37,300 @@ func (r ResourceToClone) OK() error {
+	KindNotificationEndpointHTTP:      7,
+	KindNotificationEndpointPagerDuty: 8,
+	KindNotificationEndpointSlack:     9,
+	KindNotificationRule:              10,
+	KindTask:                          11,
+	KindVariable:                      12,
+	KindDashboard:                     13,
+	KindTelegraf:                      14,
+}
+
+type exportKey struct {
+	orgID influxdb.ID
+	name  string
+	kind  Kind
+}
+
+func newExportKey(orgID influxdb.ID, k Kind, name string) exportKey {
+	return exportKey{
+		orgID: orgID,
+		name:  name,
+		kind:  k,
+	}
+}
+
+type exportCoordinator struct {
+	bucketSVC   influxdb.BucketService
+	checkSVC    influxdb.CheckService
+	dashSVC     influxdb.DashboardService
+	labelSVC    influxdb.LabelService
+	endpointSVC influxdb.NotificationEndpointService
+	ruleSVC     influxdb.NotificationRuleStore
+	taskSVC     influxdb.TaskService
+	teleSVC     influxdb.TelegrafConfigStore
+	varSVC      influxdb.VariableService
+
+	mObjects map[exportKey]Object
+}
+
+func (ex *exportCoordinator) Export(ctx context.Context, resourcesToClone []ResourceToClone, labelNames ...string) error {
+	cloneAssFn, err := ex.resourceCloneAssociationsGen(ctx, labelNames...)
+	if err != nil {
+		return err
+	}
+
+	resourcesToClone = uniqResourcesToClone(resourcesToClone)
+	sort.Slice(resourcesToClone, func(i, j int) bool {
+		iName, jName := resourcesToClone[i].Name, resourcesToClone[j].Name
+		iKind, jKind := resourcesToClone[i].Kind, resourcesToClone[j].Kind
+
+		if iKind.is(jKind) {
+			return iName < jName
+		}
+		return kindPriorities[iKind] < kindPriorities[jKind]
+	})
+
+	for _, r := range resourcesToClone {
+		err := ex.resourceCloneToKind(ctx, r, cloneAssFn)
+		if err != nil {
+			return internalErr(fmt.Errorf("failed to clone resource: resource_id=%s resource_kind=%s err=%q", r.ID, r.Kind, err))
+		}
+	}
+
+	return nil
+}
+
+func (ex *exportCoordinator) Objects() []Object {
+	objects := make([]Object, 0, len(ex.mObjects))
+	for _, obj := range ex.mObjects {
+		objects = append(objects, obj)
+	}
+
+	sort.Slice(objects, func(i, j int) bool {
+		iName, jName := objects[i].Name(), objects[j].Name()
+		iKind, jKind := objects[i].Kind, objects[j].Kind
+
+		if iKind.is(jKind) {
+			return iName < jName
+		}
+		return kindPriorities[iKind] < kindPriorities[jKind]
+	})
+
+	return objects
+}
+
+type cloneAssociationsFn func(context.Context, ResourceToClone) (associations []Resource, skipResource bool, err error)
+
+func (ex *exportCoordinator) resourceCloneToKind(ctx context.Context, r ResourceToClone, cFn cloneAssociationsFn) (e error) {
+	defer func() {
+		if e != nil {
+			e = ierrors.Wrap(e, "cloning resource")
+		}
+	}()
+
+	ass, skipResource, err := cFn(ctx, r)
+	if err != nil {
+		return err
+	}
+	if skipResource {
+		return nil
+	}
+
+	mapResource := func(orgID influxdb.ID, k Kind, object Object) {
+		if len(ass) > 0 {
+			object.Spec[fieldAssociations] = ass
+		}
+		key := newExportKey(orgID, k, object.Spec.stringShort(fieldName))
```

Comment:
quick question, where does the `fieldName` come from?

Reply:
`fieldName` is a constant that couples together the decoding and encoding fields. You can find it here in the `models.go` file 👍 

https://github.com/influxdata/influxdb/blob/52c7447333c9944e88d71b307770ce9cb1db9b2c/pkger/models.go#L777